### PR TITLE
New standard pipes

### PR DESCRIPTION
**Pull request by piloopin**
Part of #2340. Next PR will be for internationalization pipes (number, currency and date)

[Review on Reviewable](https://reviewable.io/reviews/angular/angular/2464)

### CODE REVIEW

**Comment by gdi2290** [comment]
:+1:

**Comment by tbosch** [comment]
Is this still WIP or ready to be merged?

**Comment by piloopin** [comment]
Sorry, I meant it doesn't close #2340. Updated the title.

**Comment by vicb** [comment]
@vsavkin could you please review this PR, you know Pipes better than me.

**Comment by piloopin** [comment]
@vsavkin are you looking at this PR? Just asking because I have intl pipes (date, number, currency, percent) ready to submit, but that one depends on this PR.

File: modules/angular2/src/change_detection/pipes/limit_to_pipe.ts [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+ * expression.
+ * - **a negative integer**: return _number_ items from the end of the list or string expression.
+ * - **`|limitTo|` greater than the size of the expression**: return the entire expression.
+ *
+ * When operating on a [List], the returned list is always a copy even when all
+ * the elements are being returned.
+ *
+ * # Examples
+ *
+ *     {{ 'abcdefghij' | limitTo: 4 }}       // output is 'abcd'
```
Comment (mhevery) [inline modules/angular2/src/change_detection/pipes/limit_to_pipe.ts]: this shows string examples. Could we also have a list example?

File: modules/angular2/src/change_detection/pipes/lowercase_pipe.ts [hunk 2]
```diff
@@ -46,8 +46,8 @@ export class LowerCasePipe extends Pipe {
 /**
  * @exportedAs angular2/pipes
  */
-export class LowerCaseFactory {
+export class LowerCaseFactory implements PipeFactory {
   supports(str): boolean { return isString(str); }
 
-  create(): Pipe { return new LowerCasePipe(); }
+  create(cdRef): Pipe { return new LowerCasePipe(); }
```
Comment (mhevery) [inline modules/angular2/src/change_detection/pipes/lowercase_pipe.ts]: can we add type information as much as possible. `cdRef: ChangeDetectionRef`

File: modules/angular2/src/change_detection/pipes/pipe.dart [hunk 3]
```diff
@@ -0,0 +1,37 @@
+library angular2.src.change_detection.pipes.pipe;
```
Comment (mhevery) [inline modules/angular2/src/change_detection/pipes/pipe.dart]: Why do we need a special dart file here? Can't we just generate it from the `.ts`?

File: modules/angular2/src/change_detection/pipes/uppercase_pipe.ts [hunk 4]
```diff
@@ -46,8 +46,8 @@ export class UpperCasePipe extends Pipe {
 /**
  * @exportedAs angular2/pipes
  */
-export class UpperCaseFactory {
+export class UpperCaseFactory implements PipeFactory {
   supports(str): boolean { return isString(str); }
 
-  create(): Pipe { return new UpperCasePipe(); }
+  create(cdRef): Pipe { return new UpperCasePipe(); }
```
Comment (mhevery) [inline modules/angular2/src/change_detection/pipes/uppercase_pipe.ts]: type annotation please

File: modules/angular2/src/change_detection/pipes/wrapped_value.ts [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+ *
+ * The wrapped value will be unwrapped by change detection, and the unwrapped value will be stored.
+ *
+ * @exportedAs angular2/pipes
+ */
+export class WrappedValue {
+  constructor(public wrapped: any) {}
+
+  static wrap(value: any): WrappedValue {
+    var w = _wrappedValues[_wrappedIndex++ % 5];
```
Comment (mhevery) [inline modules/angular2/src/change_detection/pipes/wrapped_value.ts]: @vsavkin why do we have a magical 5 here? What does this limit?

**Comment by mhevery** [comment]
@piloopin thanks for this. I have reviewed it and it only has minor clean ups.
@vsavkin / @tbosch Revied and ready to merge after @piloopin cleans up.

File: modules/angular2/src/change_detection/pipes/wrapped_value.ts (hunk 5 above)
Comment (vsavkin) [inline modules/angular2/src/change_detection/pipes/wrapped_value.ts]: - I think we should remove this ad-hoc pool. I added it not to allocate a new instance of wrapped value every single time we need to force change detection. Typically, you would use a single instance of WrappedValue. But I theory you could use more than one. Not sure why I picked "5" as the size of the pool.

File: modules/angular2/src/change_detection/pipes/pipe.dart (hunk 3 above)
Comment (piloopin) [inline modules/angular2/src/change_detection/pipes/pipe.dart]: This complication is a result of trying to make `transform` variable argument. This is not natively supported in Dart (which breaks the `Pipe` interface) and in javascript there was the problem of lexical this in function.apply, hence the files diverged.

I switched to a more basic straightforward of argument passing for this PR. But I believe this is not ideal and we may need some refactoring.

**Comment by piloopin** [comment]
Thanks @mhevery. Comments integrated.

**Comment by mhevery** [comment]
Reviewed 22 of 22 files at r1.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/angular/angular/2464)

**Comment by tbosch** [comment]
Please rebase ontop of master.

**Comment by piloopin** [comment]
Done.

**Comment by vsavkin** [comment]
Merged via https://github.com/angular/angular/commit/0b502588147fce6da5633898cdb357a44107394a

**Pull request closed without merging**